File: files/scripts/utils/Heap.lua

local Heap = {}
Heap.__index = Heap 

function Heap:new()
	local obj = {
		data={}
	}
	setmetatable(obj,self)
	return obj 
end

function Heap:push(priority,value)
	local d = self.data
	table.insert(d,{priority,value})
	--进行上浮操作
	local i = #d
	--当其小于
	while i>1 and d[math.floor(i/2)][1]> d[i][1]    do 
		d[math.floor(i/2)],d[i] = d[i],d[math.floor(i/2)]
		i = math.floor(i/2)
	end
end

function Heap:pop()
	local d = self.data
	if #d == 0 then return nil  end
	local top = d[1][2]
	local last = table.remove(d,#d)
	if #d>0 then
		d[1] = last
		local i = 1
		while true do
			local left = 2*i
			local right = 2*i+1
			local smallest =  i 
			if left <= #d and d[left][1] < d[smallest][1] then smallest= left end
			if right <= #d and d[right][1] < d[smallest][1] then smallest = right end   
			if smallest == i then break end
			d[smallest],d[i] = d[i],d[smallest]
			i = smallest			
		end		
	end
	return top 
end
function Heap:peek()
	return self.d[1] and self.d[1][2]	
end

function Heap:size()
	return #self.data	
end

function Heap:is_empty()
	return #self.data == 0 
end

return Heap

File: tools/algo/heap.py
"""
Min-heap priority queue (cloned from YoitaAI files/scripts/utils/Heap.lua)

Stores (priority, value) tuples.  Lower priority floats to top.
"""


class MinHeap:
    def __init__(self):
        self.data = []  # list of [priority, value]

    def push(self, priority, value):
        """Insert a (priority, value) pair and bubble up."""
        d = self.data
        d.append([priority, value])
        i = len(d) - 1
        while i > 0:
            parent = (i - 1) // 2
            if d[parent][0] > d[i][0]:
                d[parent], d[i] = d[i], d[parent]
                i = parent
            else:
                break

    def pop(self):
        """Remove and return the value with smallest priority, or None if empty."""
        d = self.data
        if not d:
            return None
        top = d[0][1]
        last = d.pop()
        if d:
            d[0] = last
            i = 0
            n = len(d)
            while True:
                left = 2 * i + 1
                right = 2 * i + 2
                smallest = i
                if left < n and d[left][0] < d[smallest][0]:
                    smallest = left
                if right < n and d[right][0] < d[smallest][0]:
                    smallest = right
                if smallest == i:
                    break
                d[smallest], d[i] = d[i], d[smallest]
                i = smallest
        return top

    def peek(self):
        """Return the value with smallest priority without removing it."""
        if self.data:
            return self.data[0][1]
        return None

    def __len__(self):
        return len(self.data)

    def is_empty(self):
        return len(self.data) == 0
